File: onyx/backend/onyx/server/query_and_chat/streaming_models.py
from enum import Enum
from typing import Annotated
from typing import Any
from typing import Literal
from typing import Union

from pydantic import BaseModel
from pydantic import Field

from onyx.context.search.models import SearchDoc
from onyx.server.query_and_chat.placement import Placement


class StreamingType(Enum):
    """Enum defining all streaming packet types. This is the single source of truth for type strings."""

    SECTION_END = "section_end"
    STOP = "stop"
    TOP_LEVEL_BRANCHING = "top_level_branching"
    ERROR = "error"

    MESSAGE_START = "message_start"
    MESSAGE_DELTA = "message_delta"
    SEARCH_TOOL_START = "search_tool_start"
    SEARCH_TOOL_QUERIES_DELTA = "search_tool_queries_delta"
    SEARCH_TOOL_DOCUMENTS_DELTA = "search_tool_documents_delta"
    OPEN_URL_START = "open_url_start"
    OPEN_URL_URLS = "open_url_urls"
    OPEN_URL_DOCUMENTS = "open_url_documents"
    IMAGE_GENERATION_START = "image_generation_start"
    IMAGE_GENERATION_HEARTBEAT = "image_generation_heartbeat"
    IMAGE_GENERATION_FINAL = "image_generation_final"
    PYTHON_TOOL_START = "python_tool_start"
    PYTHON_TOOL_DELTA = "python_tool_delta"
    CUSTOM_TOOL_START = "custom_tool_start"
    CUSTOM_TOOL_ARGS = "custom_tool_args"
    CUSTOM_TOOL_DELTA = "custom_tool_delta"
    FILE_READER_START = "file_reader_start"
    FILE_READER_RESULT = "file_reader_result"
    REASONING_START = "reasoning_start"
    REASONING_DELTA = "reasoning_delta"
    REASONING_DONE = "reasoning_done"
    CITATION_INFO = "citation_info"
    TOOL_CALL_DEBUG = "tool_call_debug"
    TOOL_CALL_ARGUMENT_DELTA = "tool_call_argument_delta"

    MEMORY_TOOL_START = "memory_tool_start"
    MEMORY_TOOL_DELTA = "memory_tool_delta"
    MEMORY_TOOL_NO_ACCESS = "memory_tool_no_access"

    DEEP_RESEARCH_PLAN_START = "deep_research_plan_start"
    DEEP_RESEARCH_PLAN_DELTA = "deep_research_plan_delta"
    RESEARCH_AGENT_START = "research_agent_start"
    INTERMEDIATE_REPORT_START = "intermediate_report_start"
    INTERMEDIATE_REPORT_DELTA = "intermediate_report_delta"
    INTERMEDIATE_REPORT_CITED_DOCS = "intermediate_report_cited_docs"


class BaseObj(BaseModel):
    type: str = ""


################################################
# Control Packets
################################################
# This one isn't strictly necessary, remove in the future
class SectionEnd(BaseObj):
    type: Literal["section_end"] = StreamingType.SECTION_END.value


class OverallStop(BaseObj):
    type: Literal["stop"] = StreamingType.STOP.value
    stop_reason: str | None = None


class TopLevelBranching(BaseObj):
    # This class is used to give advanced heads up to the frontend that the top level flow is branching
    # This is used to avoid having the frontend render the first call then rerendering the other parallel branches
    type: Literal["top_level_branching"] = StreamingType.TOP_LEVEL_BRANCHING.value

    num_parallel_branches: int


class PacketException(BaseObj):
    type: Literal["error"] = StreamingType.ERROR.value

    exception: Exception = Field(exclude=True)
    model_config = {"arbitrary_types_allowed": True}


################################################
# Reasoning Packets
################################################
# Tells the frontend to display the reasoning block
class ReasoningStart(BaseObj):
    type: Literal["reasoning_start"] = StreamingType.REASONING_START.value


# The stream of tokens for the reasoning
class ReasoningDelta(BaseObj):
    type: Literal["reasoning_delta"] = StreamingType.REASONING_DELTA.value

    reasoning: str


class ReasoningDone(BaseObj):
    type: Literal["reasoning_done"] = StreamingType.REASONING_DONE.value


################################################
# Final Agent Response Packets
################################################
# Start of the final answer
class AgentResponseStart(BaseObj):
    type: Literal["message_start"] = StreamingType.MESSAGE_START.value

    final_documents: list[SearchDoc] | None = None
    pre_answer_processing_seconds: float | None = None


# The stream of tokens for the final response
# There is no end packet for this as the stream is over and a final OverallStop packet is emitted
class AgentResponseDelta(BaseObj):
    type: Literal["message_delta"] = StreamingType.MESSAGE_DELTA.value

    content: str


# Citation info for the sidebar and inline citations
class CitationInfo(BaseObj):
    type: Literal["citation_info"] = StreamingType.CITATION_INFO.value

    # The numerical number of the citation as provided by the LLM
    citation_number: int
    # The document id of the SearchDoc (same as the field stored in the DB)
    # This is the actual document id from the connector, not the int id
    document_id: str


class ToolCallDebug(BaseObj):
    type: Literal["tool_call_debug"] = StreamingType.TOOL_CALL_DEBUG.value

    tool_call_id: str
    tool_name: str
    tool_args: dict[str, Any]


################################################
# Tool Packets
################################################
# Search tool is called and the UI block needs to start
class SearchToolStart(BaseObj):
    type: Literal["search_tool_start"] = StreamingType.SEARCH_TOOL_START.value

    is_internet_search: bool = False


# Queries coming through as the LLM determines what to search
# Mostly for query expansions and advanced search strategies
class SearchToolQueriesDelta(BaseObj):
    type: Literal["search_tool_queries_delta"] = (
        StreamingType.SEARCH_TOOL_QUERIES_DELTA.value
    )

    queries: list[str]


# Documents coming through as the system knows what to add to the context
class SearchToolDocumentsDelta(BaseObj):
    type: Literal["search_tool_documents_delta"] = (
        StreamingType.SEARCH_TOOL_DOCUMENTS_DELTA.value
    )

    # This cannot be the SavedSearchDoc as this is yielded by the SearchTool directly
    # which does not save documents to the DB.
    documents: list[SearchDoc]


# OpenURL tool packets - 3-stage sequence
class OpenUrlStart(BaseObj):
    """Signal that OpenURL tool has started."""

    type: Literal["open_url_start"] = StreamingType.OPEN_URL_START.value


class OpenUrlUrls(BaseObj):
    """URLs to be fetched (sent before crawling begins)."""

    type: Literal["open_url_urls"] = StreamingType.OPEN_URL_URLS.value

    urls: list[str]


class OpenUrlDocuments(BaseObj):
    """Final documents after crawling completes."""

    type: Literal["open_url_documents"] = StreamingType.OPEN_URL_DOCUMENTS.value

    documents: list[SearchDoc]


# Image generation starting, needs to allocate a placeholder block for it on the UI
class ImageGenerationToolStart(BaseObj):
    type: Literal["image_generation_start"] = StreamingType.IMAGE_GENERATION_START.value


# Since image generation can take a while
# we send a heartbeat to the frontend to keep the UI/connection alive
class ImageGenerationToolHeartbeat(BaseObj):
    type: Literal["image_generation_heartbeat"] = (
        StreamingType.IMAGE_GENERATION_HEARTBEAT.value
    )


# Represents an image generated by an image generation tool
class GeneratedImage(BaseModel):
    """Represents an image generated by an image generation tool."""

    file_id: str
    url: str
    revised_prompt: str
    shape: str | None = None


# The final generated images all at once at the end of image generation
class ImageGenerationFinal(BaseObj):
    type: Literal["image_generation_final"] = StreamingType.IMAGE_GENERATION_FINAL.value

    images: list[GeneratedImage]


class PythonToolStart(BaseObj):
    type: Literal["python_tool_start"] = StreamingType.PYTHON_TOOL_START.value
    code: str


class PythonToolDelta(BaseObj):
    type: Literal["python_tool_delta"] = StreamingType.PYTHON_TOOL_DELTA.value

    stdout: str = ""
    stderr: str = ""
    file_ids: list[str] = []


# Custom tool being called, first allocate a placeholder block for it on the UI
class CustomToolStart(BaseObj):
    type: Literal["custom_tool_start"] = StreamingType.CUSTOM_TOOL_START.value

    tool_name: str
    tool_id: int | None = None


class CustomToolArgs(BaseObj):
    type: Literal["custom_tool_args"] = StreamingType.CUSTOM_TOOL_ARGS.value

    tool_name: str
    tool_args: dict[str, Any]


class CustomToolErrorInfo(BaseModel):
    is_auth_error: bool = False
    status_code: int
    message: str


# The allowed streamed packets for a custom tool
class CustomToolDelta(BaseObj):
    type: Literal["custom_tool_delta"] = StreamingType.CUSTOM_TOOL_DELTA.value

    tool_name: str
    tool_id: int | None = None
    response_type: str
    # For non-file responses
    data: dict | list | str | int | float | bool | None = None
    # For file-based responses like image/csv
    file_ids: list[str] | None = None
    error: CustomToolErrorInfo | None = None


class ToolCallArgumentDelta(BaseObj):
    type: Literal["tool_call_argument_delta"] = (
        StreamingType.TOOL_CALL_ARGUMENT_DELTA.value
    )

    tool_type: str
    argument_deltas: dict[str, Any]


################################################
# File Reader Packets
################################################
class FileReaderStart(BaseObj):
    type: Literal["file_reader_start"] = StreamingType.FILE_READER_START.value


class FileReaderResult(BaseObj):
    type: Literal["file_reader_result"] = StreamingType.FILE_READER_RESULT.value

    file_name: str
    file_id: str
    start_char: int
    end_char: int
    total_chars: int
    # Short previews of the retrieved text for the collapsed/expanded UI
    preview_start: str = ""
    preview_end: str = ""


# Memory Tool Packets
################################################
class MemoryToolStart(BaseObj):
    type: Literal["memory_tool_start"] = StreamingType.MEMORY_TOOL_START.value


class MemoryToolDelta(BaseObj):
    type: Literal["memory_tool_delta"] = StreamingType.MEMORY_TOOL_DELTA.value

    memory_text: str
    operation: Literal["add", "update"]
    memory_id: int | None = None
    index: int | None = None


class MemoryToolNoAccess(BaseObj):
    type: Literal["memory_tool_no_access"] = StreamingType.MEMORY_TOOL_NO_ACCESS.value


################################################
# Deep Research Packets
################################################
class DeepResearchPlanStart(BaseObj):
    type: Literal["deep_research_plan_start"] = (
        StreamingType.DEEP_RESEARCH_PLAN_START.value
    )


class DeepResearchPlanDelta(BaseObj):
    type: Literal["deep_research_plan_delta"] = (
        StreamingType.DEEP_RESEARCH_PLAN_DELTA.value
    )

    content: str


class ResearchAgentStart(BaseObj):
    type: Literal["research_agent_start"] = StreamingType.RESEARCH_AGENT_START.value
    research_task: str


class IntermediateReportStart(BaseObj):
    type: Literal["intermediate_report_start"] = (
        StreamingType.INTERMEDIATE_REPORT_START.value
    )


class IntermediateReportDelta(BaseObj):
    type: Literal["intermediate_report_delta"] = (
        StreamingType.INTERMEDIATE_REPORT_DELTA.value
    )
    content: str


class IntermediateReportCitedDocs(BaseObj):
    type: Literal["intermediate_report_cited_docs"] = (
        StreamingType.INTERMEDIATE_REPORT_CITED_DOCS.value
    )
    cited_docs: list[SearchDoc] | None = None


################################################
# Packet Object
################################################
# Discriminated union of all possible packet object types
PacketObj = Union[
    # Control Packets
    OverallStop,
    SectionEnd,
    TopLevelBranching,
    PacketException,
    # Agent Response Packets
    AgentResponseStart,
    AgentResponseDelta,
    # Tool Packets
    SearchToolStart,
    SearchToolQueriesDelta,
    SearchToolDocumentsDelta,
    ImageGenerationToolStart,
    ImageGenerationToolHeartbeat,
    ImageGenerationFinal,
    OpenUrlStart,
    OpenUrlUrls,
    OpenUrlDocuments,
    PythonToolStart,
    PythonToolDelta,
    CustomToolStart,
    CustomToolArgs,
    CustomToolDelta,
    FileReaderStart,
    FileReaderResult,
    MemoryToolStart,
    MemoryToolDelta,
    MemoryToolNoAccess,
    # Reasoning Packets
    ReasoningStart,
    ReasoningDelta,
    ReasoningDone,
    # Citation Packets
    CitationInfo,
    ToolCallDebug,
    ToolCallArgumentDelta,
    # Deep Research Packets
    DeepResearchPlanStart,
    DeepResearchPlanDelta,
    ResearchAgentStart,
    IntermediateReportStart,
    IntermediateReportDelta,
    IntermediateReportCitedDocs,
]


class Packet(BaseModel):
    placement: Placement

    obj: Annotated[PacketObj, Field(discriminator="type")]

File: onyx/web/src/app/app/services/streamingModels.ts
import { OnyxDocument } from "@/lib/search/interfaces";

// Base interface for all streaming objects
interface BaseObj {
  type: string;
}

export enum PacketType {
  MESSAGE_START = "message_start",
  MESSAGE_DELTA = "message_delta",
  MESSAGE_END = "message_end",

  STOP = "stop",
  SECTION_END = "section_end",
  TOP_LEVEL_BRANCHING = "top_level_branching",
  ERROR = "error",

  // Specific tool packets
  SEARCH_TOOL_START = "search_tool_start",
  SEARCH_TOOL_QUERIES_DELTA = "search_tool_queries_delta",
  SEARCH_TOOL_DOCUMENTS_DELTA = "search_tool_documents_delta",
  IMAGE_GENERATION_TOOL_START = "image_generation_start",
  IMAGE_GENERATION_TOOL_DELTA = "image_generation_final",
  PYTHON_TOOL_START = "python_tool_start",
  PYTHON_TOOL_DELTA = "python_tool_delta",
  FETCH_TOOL_START = "open_url_start",
  FETCH_TOOL_URLS = "open_url_urls",
  FETCH_TOOL_DOCUMENTS = "open_url_documents",

  // Tool call argument delta (streams tool args before tool executes)
  TOOL_CALL_ARGUMENT_DELTA = "tool_call_argument_delta",

  // Custom tool packets
  CUSTOM_TOOL_START = "custom_tool_start",
  CUSTOM_TOOL_ARGS = "custom_tool_args",
  CUSTOM_TOOL_DELTA = "custom_tool_delta",

  // File reader tool packets
  FILE_READER_START = "file_reader_start",
  FILE_READER_RESULT = "file_reader_result",
  // Memory tool packets
  MEMORY_TOOL_START = "memory_tool_start",
  MEMORY_TOOL_DELTA = "memory_tool_delta",
  MEMORY_TOOL_NO_ACCESS = "memory_tool_no_access",

  // Reasoning packets
  REASONING_START = "reasoning_start",
  REASONING_DELTA = "reasoning_delta",
  REASONING_DONE = "reasoning_done",

  // Citation packets
  CITATION_START = "citation_start",
  CITATION_END = "citation_end",
  // Backend sends individual citation_info packets during streaming
  CITATION_INFO = "citation_info",

  // Deep Research packets
  DEEP_RESEARCH_PLAN_START = "deep_research_plan_start",
  DEEP_RESEARCH_PLAN_DELTA = "deep_research_plan_delta",
  RESEARCH_AGENT_START = "research_agent_start",
  INTERMEDIATE_REPORT_START = "intermediate_report_start",
  INTERMEDIATE_REPORT_DELTA = "intermediate_report_delta",
  INTERMEDIATE_REPORT_CITED_DOCS = "intermediate_report_cited_docs",
}

export const CODE_INTERPRETER_TOOL_TYPES = {
  PYTHON: "python",
} as const;

// Basic Message Packets
export interface MessageStart extends BaseObj {
  id: string;
  type: "message_start";
  content: string;

  final_documents: OnyxDocument[] | null;
  pre_answer_processing_seconds?: number;
}

export interface MessageDelta extends BaseObj {
  content: string;
  type: "message_delta";
}

export interface MessageEnd extends BaseObj {
  type: "message_end";
}

// Control Packets
export enum StopReason {
  FINISHED = "finished",
  USER_CANCELLED = "user_cancelled",
}

export interface Stop extends BaseObj {
  type: "stop";
  stop_reason?: StopReason;
}

export interface SectionEnd extends BaseObj {
  type: "section_end";
}

export interface TopLevelBranching extends BaseObj {
  type: "top_level_branching";
  num_parallel_branches: number;
}

export interface PacketError extends BaseObj {
  type: "error";
  message?: string;
}

// Specific tool packets
export interface SearchToolStart extends BaseObj {
  type: "search_tool_start";
  is_internet_search?: boolean;
}

export interface SearchToolQueriesDelta extends BaseObj {
  type: "search_tool_queries_delta";
  queries: string[];
}

export interface SearchToolDocumentsDelta extends BaseObj {
  type: "search_tool_documents_delta";
  documents: OnyxDocument[];
}

export type ImageShape = "square" | "landscape" | "portrait";

interface GeneratedImage {
  file_id: string;
  url: string;
  revised_prompt: string;
  shape?: ImageShape;
}

export interface ImageGenerationToolStart extends BaseObj {
  type: "image_generation_start";
}

export interface ImageGenerationToolDelta extends BaseObj {
  type: "image_generation_final";
  images: GeneratedImage[];
}

export interface PythonToolStart extends BaseObj {
  type: "python_tool_start";
  code: string;
}

export interface PythonToolDelta extends BaseObj {
  type: "python_tool_delta";
  stdout: string;
  stderr: string;
  file_ids: string[];
}

export interface ToolCallArgumentDelta extends BaseObj {
  type: "tool_call_argument_delta";
  tool_type: string;
  tool_id: string;
  argument_deltas: Record<string, unknown>;
}

export interface FetchToolStart extends BaseObj {
  type: "open_url_start";
}

export interface FetchToolUrls extends BaseObj {
  type: "open_url_urls";
  urls: string[];
}

export interface FetchToolDocuments extends BaseObj {
  type: "open_url_documents";
  documents: OnyxDocument[];
}

// Custom Tool Packets
export interface CustomToolErrorInfo {
  is_auth_error: boolean;
  status_code: number;
  message: string;
}

export interface CustomToolStart extends BaseObj {
  type: "custom_tool_start";
  tool_name: string;
  tool_id?: number | null;
}

export interface CustomToolArgs extends BaseObj {
  type: "custom_tool_args";
  tool_name: string;
  tool_args: Record<string, any>;
}

export interface CustomToolDelta extends BaseObj {
  type: "custom_tool_delta";
  tool_name: string;
  tool_id?: number | null;
  response_type: string;
  data?: any;
  file_ids?: string[] | null;
  error?: CustomToolErrorInfo | null;
}

// File Reader Packets
export interface FileReaderStart extends BaseObj {
  type: "file_reader_start";
}

export interface FileReaderResult extends BaseObj {
  type: "file_reader_result";
  file_name: string;
  file_id: string;
  start_char: number;
  end_char: number;
  total_chars: number;
  preview_start: string;
  preview_end: string;
}
// Memory Tool Packets
export interface MemoryToolStart extends BaseObj {
  type: "memory_tool_start";
}

export interface MemoryToolDelta extends BaseObj {
  type: "memory_tool_delta";
  memory_text: string;
  operation: "add" | "update";
  memory_id: number | null;
  index: number | null;
}

export interface MemoryToolNoAccess extends BaseObj {
  type: "memory_tool_no_access";
}

// Reasoning Packets
export interface ReasoningStart extends BaseObj {
  type: "reasoning_start";
}

export interface ReasoningDelta extends BaseObj {
  type: "reasoning_delta";
  reasoning: string;
}

export interface ReasoningDone extends BaseObj {
  type: "reasoning_done";
}

// Citation Packets
export interface StreamingCitation {
  citation_num: number;
  document_id: string;
}

export interface CitationStart extends BaseObj {
  type: "citation_start";
}

// Individual citation info packet (sent during streaming from backend)
export interface CitationInfo extends BaseObj {
  type: "citation_info";
  citation_number: number;
  document_id: string;
}

// Deep Research Plan Packets
export interface DeepResearchPlanStart extends BaseObj {
  type: "deep_research_plan_start";
}

export interface DeepResearchPlanDelta extends BaseObj {
  type: "deep_research_plan_delta";
  content: string;
}

export interface ResearchAgentStart extends BaseObj {
  type: "research_agent_start";
  research_task: string;
}

export interface IntermediateReportStart extends BaseObj {
  type: "intermediate_report_start";
}

export interface IntermediateReportDelta extends BaseObj {
  type: "intermediate_report_delta";
  content: string;
}

export interface IntermediateReportCitedDocs extends BaseObj {
  type: "intermediate_report_cited_docs";
  cited_docs: OnyxDocument[] | null;
}

export type ChatObj = MessageStart | MessageDelta | MessageEnd;

export type StopObj = Stop;

export type SectionEndObj = SectionEnd;

export type TopLevelBranchingObj = TopLevelBranching;

export type PacketErrorObj = PacketError;

// Specific tool objects
export type SearchToolObj =
  | SearchToolStart
  | SearchToolQueriesDelta
  | SearchToolDocumentsDelta
  | SectionEnd
  | PacketError;
export type ImageGenerationToolObj =
  | ImageGenerationToolStart
  | ImageGenerationToolDelta
  | SectionEnd
  | PacketError;
export type PythonToolObj =
  | PythonToolStart
  | PythonToolDelta
  | ToolCallArgumentDelta
  | SectionEnd
  | PacketError;
export type FetchToolObj =
  | FetchToolStart
  | FetchToolUrls
  | FetchToolDocuments
  | SectionEnd
  | PacketError;
export type CustomToolObj =
  | CustomToolStart
  | CustomToolArgs
  | CustomToolDelta
  | SectionEnd
  | PacketError;
export type FileReaderToolObj =
  | FileReaderStart
  | FileReaderResult
  | SectionEnd
  | PacketError;
export type MemoryToolObj =
  | MemoryToolStart
  | MemoryToolDelta
  | MemoryToolNoAccess
  | SectionEnd
  | PacketError;
export type NewToolObj =
  | SearchToolObj
  | ImageGenerationToolObj
  | PythonToolObj
  | FetchToolObj
  | CustomToolObj
  | FileReaderToolObj
  | MemoryToolObj;

export type ReasoningObj =
  | ReasoningStart
  | ReasoningDelta
  | ReasoningDone
  | SectionEnd
  | PacketError;

export type CitationObj =
  | CitationStart
  | CitationInfo
  | SectionEnd
  | PacketError;

export type DeepResearchPlanObj =
  | DeepResearchPlanStart
  | DeepResearchPlanDelta
  | SectionEnd;

export type ResearchAgentObj =
  | ResearchAgentStart
  | IntermediateReportStart
  | IntermediateReportDelta
  | IntermediateReportCitedDocs
  | SectionEnd;

// Union type for all possible streaming objects
export type ObjTypes =
  | ChatObj
  | NewToolObj
  | ReasoningObj
  | StopObj
  | SectionEndObj
  | TopLevelBranchingObj
  | CitationObj
  | DeepResearchPlanObj
  | ResearchAgentObj
  | PacketErrorObj
  | CitationObj;

// Placement interface for packet positioning
export interface Placement {
  turn_index: number;
  tab_index?: number; // For parallel tool calls - tools with same turn_index but different tab_index run in parallel
  sub_turn_index?: number | null;
  model_index?: number | null; // For multi-model answer generation - identifies which model produced this packet
}

// Packet wrapper for streaming objects
export interface Packet {
  placement: Placement;
  obj: ObjTypes;
}

export interface ChatPacket {
  placement: Placement;
  obj: ChatObj;
}

export interface StopPacket {
  placement: Placement;
  obj: StopObj;
}

export interface CitationPacket {
  placement: Placement;
  obj: CitationObj;
}

// New specific tool packet types
export interface SearchToolPacket {
  placement: Placement;
  obj: SearchToolObj;
}

export interface ImageGenerationToolPacket {
  placement: Placement;
  obj: ImageGenerationToolObj;
}

export interface PythonToolPacket {
  placement: Placement;
  obj: PythonToolObj;
}

export interface FetchToolPacket {
  placement: Placement;
  obj: FetchToolObj;
}

export interface CustomToolPacket {
  placement: Placement;
  obj: CustomToolObj;
}

export interface FileReaderToolPacket {
  placement: Placement;
  obj: FileReaderToolObj;
}
export interface MemoryToolPacket {
  placement: Placement;
  obj: MemoryToolObj;
}

export interface ReasoningPacket {
  placement: Placement;
  obj: ReasoningObj;
}

export interface SectionEndPacket {
  placement: Placement;
  obj: SectionEndObj;
}

export interface TopLevelBranchingPacket {
  placement: Placement;
  obj: TopLevelBranchingObj;
}

export interface DeepResearchPlanPacket {
  placement: Placement;
  obj: DeepResearchPlanObj;
}

export interface ResearchAgentPacket {
  placement: Placement;
  obj: ResearchAgentObj;
}
